File: Harness Python/providers/openai_compat.py
import asyncio
from openai import AsyncOpenAI
from providers.base import BaseProvider
from tenacity import retry, stop_after_attempt, wait_exponential

class OpenAICompatProvider(BaseProvider):
    def __init__(self, base_url: str, api_key: str, model: str, provider_name: str, timeout: int = 30):
        self.base_url = base_url
        self.api_key = api_key
        self.model = model
        self.name = provider_name
        self.timeout = timeout
        self.client = AsyncOpenAI(api_key=api_key, base_url=base_url, timeout=timeout)

    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=2, max=10),
        reraise=True
    )
    async def complete(self, messages: list[dict], **kwargs) -> str:
        resp = await self.client.chat.completions.create(
            model=self.model,
            messages=messages,
            max_tokens=kwargs.get("max_tokens", 1024)
        )
        return resp.choices[0].message.content

    async def health_check(self) -> bool:
        try:
            await self.complete([{"role": "user", "content": "Hi"}], max_tokens=5)
            return True
        except Exception as e:
            print(f"[{self.name}] Health check failed: {e}")
            return False

File: Harness Typescript/src/providers/openai-compat.ts
import OpenAI from 'openai';
import pRetry from 'p-retry';
import { BaseProvider } from './base.js';

export class OpenAICompatProvider extends BaseProvider {
  private client: OpenAI;
  private timeout: number;

  constructor(apiKey: string, baseUrl: string, model: string, providerName: string, timeout: number = 30000) {
    super(providerName, model);
    this.timeout = timeout;

    this.client = new OpenAI({
      apiKey: apiKey,
      baseURL: baseUrl,
      timeout: this.timeout,
    });
  }

  async complete(messages: any[], options: any = {}): Promise<string> {
    const runComplete = async () => {
      const resp = await this.client.chat.completions.create({
        model: this.model,
        messages: messages,
        max_tokens: options.max_tokens || 1024,
      });

      return resp.choices[0].message?.content || "";
    };

    return pRetry(runComplete, {
      retries: 3,
      minTimeout: 2000,
      maxTimeout: 10000,
    });
  }

  async healthCheck(): Promise<boolean> {
    try {
      await this.complete([{ role: "user", content: "Hi" }], { max_tokens: 5 });
      return true;
    } catch (e) {
      console.log(`[${this.name}] Health check failed:`, (e as Error).message);
      return false;
    }
  }
}
